[NAS] 미션 5: 보드에 위치 부여 및 점수 계산

Malloc72P opened the pull request:
안녕하세요!
백엔드 클래스의 NAS예요! 😄 

미션5에 대한 Pull Request를 남기게 되었습니다! 🚢 

작업내용은 아래와 같습니다!

### 1. 기물의 색과 종류에 따른 enum을 적용했습니다.
색은 ColorOfChessPiece, 종류는 NameOfChessPiece라는 enum으로 구현했습니다.
또한 기물의 종류에 따른 P, K등의 식별문자를 NameOfChessPiece enum으로 관리했습니다.
추가로, 기물의 점수도 해당 enum에서 관리하도록 했습니다.
```java
public enum NameOfChessPiece {
    PAWN("P", 1),
    ROOK("R", 5),
/*~~~*/
```
### 2. 기물과 색에 해당하는 기물의 개수를 반환하는 기능을 구현했습니다.
`getNumberOfDesignatedPiece()`메서드를 구현해서, 해당 기능을 구현했습니다.
패러미터로 기물의 이름과 색상을 받아서, 지정된 색상에 해당하는 체스말의 개수를 구해서 반환하게끔 하였습니다.

### 3. 주어진 위치의 기물을 조회하는 기능을 구현했습니다.
해당 기능의 구현을 위해서, `ChessCoordinate`클래스를 만들었습니다.
`ChessCoordinate`클래스는 좌표관리를 담당하는 객체의 클래스입니다.
해당 클래스에는 `parseStringCoordinate(String strCoordinate)`라는 정적 메서드가 있습니다.
`ChessCoordinate`클래스는 `e5`와 같은 문자열 좌표를, 체스보드의 배열상에서의 좌표로 변환하는 서비스를 제공합니다. 이 서비스는 Board클래스에서 이용합니다.
위의 기능을 이용해서 주어진 위치가 배열에서 어디인지 알아내고, 해당 위치의 체스말을 반환하도록 구현하였습니다.

### 4. 임의의 기물을 체스판에 추가하는 기능을 구현했습니다.
체스판의 어디에 추가할지는 문자열로 입력받습니다. 따라서 체스말을 찾는 기능을 구현할때와 마찬가지로, `ChessCoordinate`클래스의 도움을 받아서 어디에 추가해야 하는지를 알아냅니다.
또한, 예전에 `public void add(ChessPiece piece, int rankIdx, int fileIdx)`라는 메서드를 구현했습니다. 이 메서드는 배열상의 좌표를 가지고 체스판에 체스말을 추가하는 메서드입니다. 그래서 이번 기능은 오버로딩을 이용해서, 간단하게 아래와 같이 구현할 수 있었습니다.
```java
public void add(ChessPiece piece, String strCoordinate) {
    ChessCoordinate coordinate = parseStringCoordinate(strCoordinate);
    add(piece, coordinate.rankIndex, coordinate.fileIndex);
}
```

### 5. 점수계산 기능을 구현했습니다
NameOfChessPiece enum에 추가한 point필드를 이용해서 구현했습니다. 


### 6. 점수가 높은 순으로 정렬된 체스말을 컬렉션에 담아 리턴하는 기능을 구현하였습니다
체스말 클래스인 `ChessPiece`클래스에 두개의 `Comparator`를 정의하였습니다. 해당 `Comparator`는 체스말의 점수를 가지고 비교합니다. 이렇게 구현된 `Comparator`는 Board클래스의 `getPiecesOfColor`메서드에서 사용합니다. 컬렉션에 담은 체스말을 오름차순 또는 내림차순으로 정렬하기 위해서입니다.

### 7. 코드를 좀 더 개선했습니다
* if문을 쓸때, 한줄짜리 코드인 경우, 블록안에 넣지 않고 썼었는데, 전부 블록안에 넣어서 쓰도록 수정했습니다. 확실히, 이렇게 짜니까 보기도 더 좋고, 나중에 실수할 확률이 줄어들 것 같습니다.
* 배열을 다루는 for루프에서 0부터 시작하도록 고쳤습니다. 예전엔 실제세계에서 체스판을 다루는거랑 같게 하려고 1부터 시작했었습니다. 대신 내부적으로는 -1을 해주었는데, 이게 오히려 코드를 복잡하게 만드는 것 같아서 0부터 시작하도록 했습니다.

이상으로 글을 마치겠습니다! 😆 
다시 날이 추워지는 것 같은데, 감기 조심하세요! ❄️

Malloc72P committed:
if문을 블록안에 넣어라

- if문은 되도록이면 블록안에 작성하는게 실수를 피하는 좋은 방법이라고 한다.
- 그래서 블록안에 if문을 작성하지 않은 모든 코드를 수정함.

Malloc72P committed:
배열을 다루는 for루프의 인덱스가 0부터 시작하도록 수정하라

- 배열을 다루는 for 루프는 가능하면 0부터 시작하는것이 사이드 이펙트를 줄이는데 도움이 된다고 한다.
- 그래서 Board클래스에서 체스보드를 구현할 때, for루프가 0부터 시작하도록 수정함

Malloc72P committed:
상수값을 BLANK_1,2,3,4처럼 쓰는것 보다 나은 방법을 찾고 적용하라

- 상수를 4개 연달아 선언하기보다, 배열로 선언하는 방법을 택하여 중복코드를 줄임.

Malloc72P committed:
ChessPiece의 name, color는 이름이 너무 장황하니, 명확한 이름으로 수정하라

- 각각 nameOfChessPiece, colorOfChessPiece로 수정해서 명확한 이름을 갖도록 수정함

Malloc72P committed:
ChessPiece클래스의 getRepresentation메서드의 코드길이를 줄여라

- 코드가 너무 길어서 조금 짧게 만들고자 함.
- 문자열을 넘기는 대신, NameOfChessPiece enum을 넘기고,
  거기서 getRepresentation()메서드를 쓰도록 수정함

Malloc72P committed:
ChessPiece클래스의 getRepresentation메서드 수정으로 인한 오류를 수정하라

- 메서드가 수정되었으나, 수정 전의 방법으로 메서드를 사용하던 코드가 있어서 오류가 발생하였음.
- 따라서, 해당 오류를 수정함

Malloc72P committed:
ChessPieceTest와 BoardTest에서의 오타를 수정하라

Malloc72P committed:
Enum에서 상수 선언라인과 멤버변수 선언 사이에 공백라인을 추가하라

Malloc72P committed:
if, else문으로 구성된 복잡한 테스트코드를 assertJ를 이용하도록 수정하라

Malloc72P committed:
외부에서 접근하지 않는 NEWLINE상수를 private으로 수정하라

Malloc72P committed:
NameOfChessPiece enum에 NO_PIECE상수를 추가하라

- ChessPiece의 createBlankPiece메서드에서 PAWN대신 NO_PIECE를 쓰도록 수정함

Malloc72P committed:
ChessPieceTest에서의 팩토리 메서드 테스트코드의 중복을 줄여라

- enum을 활용해서 중복코드를 제거하도록 메서드를 추출함

Malloc72P committed:
기물과 색에 해당하는 기물의 개수를 반환하는 메서드를 작성하라

- Board클래스에 getNumberOfDesignatedPiece()메서드를 작성해서, 기물의 개수를 반환하는 기능을 구현함
- 해당 기능 테스트를 위해서, 문자열 배열을 받아서 체스판을 초기화하는 기능을 구현함. : initBoardByStringArray

Malloc72P committed:
Board클래스에 주어진 위치의 기물을 조회하는 기능을 구현하라

- e6와 같은 문자열 좌표를 읽고, ChessCoordinate클래스로 변환하는 기능을 구현함.
- ChessCoordinate클래스의 생성자는 같은 패키지에서만 생성할 수 있도록, protected생성자로 선언함.
  이는 같은 패키지에 있는 Board클래스에서만 생성할 수 있게 하기 위함임.
- 주어진 위치의 기물을 조회하는 기능은 finePiece메서드로 구현함.
- 기존에 findPawn메서드가 있었는데, 이를 findPiece로 이름을 바꾸고, 메서드 오버로딩을 통해, 문자열 좌표로 찾는 기능을 구현함.

Malloc72P committed:
ChessPiece클래스에 같은 속성인지 검사하는 메서드를 추가하라

- getNumberOfDesignatedPiece메서드에서 같은 이름과 색인지 검사하기 위해서 불필요하게 문자열로 변환하는 과정이 있었음.
- 해당 기능이 불필요하다고 판단하여, ChessPiece클래스에 hasSameProperty라는 메서드를 추가하고,
  getNumberOfDesignatedPiece메서드에서 쓰도록 수정함
- 추가로 initRankByString메서드에서의 색깔을 검사하는 코드를 좀 더 꼼꼼히 검사하도록 수정함

Malloc72P committed:
문자열 좌표로 체스판 위에 체스말을 추가하는 기능을 구현하라

- add()메서드를 오버로딩해서, 문자열 좌표로 체스판 위에 체스말을 추가하도록 구현함.
- 문자열을 읽어서 ChessCoordinate객체로 변환하는 메서드를 이용해서 배열상 좌표를 획득하도록 구현함
- 획득한 좌표로 기존의 add()메서드를 사용하도록 구현함.

Malloc72P committed:
Board클래스에서 좌표값을 처리하는 책임을 덜어주어라

- 문자열로 구성된 좌표를 ChessCoordinate객체로 변환하는 책임을 ChessCoordinate 클래스로 옮겨줌.
- 추가로 모든 문자열 좌표를 적절하게 변환할 수 있는지 테스트하는 코드를 작성하고 테스트하였음.

Malloc72P committed:
Board의 점수를 계산하는 기능을 구현하라

- 각각의 체스말에 점수를 부여하기 위해 NameOfChessPiece Enum에 점수를 표현하는 필드를 추가함
- Board클래스에 calculatePoint메서드를 추가하여 점수계산기능을 구현함

Malloc72P committed:
Board클래스에 체스말의 점수가 높은 순으로 정렬된 체스말컬렉션을 반환하는 기능을 구현하라

- getPiecesOfColor메서드를 구현하여 해당 기능을 구현함
- 패러미터로 어떤 색상의 체스말 목록을 구할지 결정함. 두번째 패러미터로 오름차순, 내림차순으로 정렬할지 결정하도록 구현함
- ChessPiece클래스에 comparatorAsc, Desc를 구현함. 이를 이용해서 정렬하도록 구현함

wheejuni commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -108,34 +99,136 @@ private String getRepresentationOfRank(int rankIdx) {
-        if (isInvalidIdx(rankIdx) || isInvalidIdx(fileIdx))
-            throw new InvalidParameterException("index exceeded the bounds of the Board");
-        return chessCells[rankIdx - 1][fileIdx - 1];
+    public void add(ChessPiece piece, int rankIdx, int fileIdx) {
+        if (piece == null) {
+            throw new InvalidParameterException("Null value cannot be added in Board");
+        }
+        if (isInvalidIdx(rankIdx) || isInvalidIdx(fileIdx)) {
+            throw new InvalidParameterException("index exceeded the bounds of the Board. rankIndex : " + rankIdx + " fileIndex : " + fileIdx);
+        }
+        chessCells[rankIdx][fileIdx] = piece;
+        if (!piece.isBlank()) {
+            numberOfPieces++;
+        }
```
여기서 사용된 `idx` 라는 축약어는 모두 완전한 단어로 바꿔주세요.

Malloc72P replied:
넵! 알겠습니다. 축약어보다는 완전한 단어가 더 좋을 것 같네요!
앞으로는 축약어 사용을 피하는 편이 좋을 것 같다는 생각이 드네요 😄 
해당 내용 수정해서 커밋하겠습니다!

wheejuni commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,94 +1,85 @@
+    public static final char END_OF_FILE = 'h';
 
     private final ChessPiece[][] chessCells;
     private int numberOfPieces = 0;
```
이 변수를 없애는 리팩토링을 진행하면 좋을 것 같은데요. 
설계에 조금 변경이 필요할 것 같기는 합니다.

wheejuni replied:
아... 이 변수란 `numberOfPieces` 를 말합니다.

Malloc72P replied:
음 혹시 어떤 변수를 말씀하시는건지 알 수 있을까요?
chessCells변수는 체스보드에 배치된 체스말을 표현하는 변수라서 꼭 필요합니다.

numberOfPieces변수도 보드에 배치된 체스말의 개수를 유지하는데 필요한 변수라서
해당 변수도 꼭 필요해요. 혹시 어떤 이유로 없애는 편이 더 나은지 알 수 있을까요?

Malloc72P replied:
`numberOfPieces`변수를 만든 이유는, 체스보드에 배치된 말의 개수를 빠르게 획득하기 위해서입니다. 체스판에 말이 몇개 배치되었는지 알고싶을때마다 2차원 배열을 전부 순회하는건 너무 비효율적이라고 생각했어요.

그런데 어차피 Board클래스의 `chessCells`는 private이고, 여기에 객체를 추가하거나 없애는 연산은 오직 Board클래스에만 있으니까, add 또는 remove할때 numberOfPieces를 업데이트해주면, 나중에 체스말이 몇개있는지 알고 싶을때, 2차원 배열을 전체 순회할 필요 없이 `numberOfPieces`변수값만 한번 읽어서 리턴하면 되니까 더 좋을 것 같아서 작성하게 되었어요.

음.. 확실히 이렇게 하면 프로그램의 복잡도가 증가할 거라고 생각해요. 솔직히 2차원배열이라고 해봤자 8 X 8 크기밖에 안돼서 그냥 전체순회해도 문제는 없을 것 같기는 한데, 저는 그래도 numberOfPieces변수를 유지하는편이 효율성 측면에서 더 낫지 않을까 하는 생각이 드네요.

그런데 다시 생각해보니까 그냥 numberOfPieces변수를 없애는 편이 나을 것 같아요 ㅎㅎ;;
이거 그대로 두면 나중에 큰 실수 할 것 같네요 ㅋㅋ 😆 
이 부분도 수정해서 커밋하겠습니다!

wheejuni commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -108,34 +99,136 @@ private String getRepresentationOfRank(int rankIdx) {
+    }
+
+    public float calculatePoint(ColorOfChessPiece color) {
+        float totalPoint = 0;
```
외부에 변수가 선언되지 않고도, `for loop`나 `stream`을 이용해서 진행해볼 방법은 없을까요?

Malloc72P replied:
`외부에 변수를 선언하지 않고도` 라는 말을 잘 이해하지 못하겠어요.
코멘트를 보면 `float totalPoint = 0`을 말씀하신 것 같은데요, 이 변수를 선언하고 사용하는데에 어떤 문제가 있을까요?

wheejuni commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -108,34 +99,136 @@ private String getRepresentationOfRank(int rankIdx) {
+
+    public float calculatePoint(ColorOfChessPiece color) {
+        float totalPoint = 0;
+        for (int fileIndex = 0; fileIndex < LENGTH_OF_BOARD; fileIndex++) {
+            int numOfPawnInFile = 0;
+            for (int rankIndex = 0; rankIndex < LENGTH_OF_BOARD; rankIndex++) {
+                ChessPiece currentPiece = chessCells[rankIndex][fileIndex];
+                if (currentPiece.getColorOfChessPiece() != color) {
+                    continue;
+                }
+                NameOfChessPiece nameOfCurrentPiece = currentPiece.getNameOfChessPiece();
+                if (nameOfCurrentPiece == NameOfChessPiece.NO_PIECE) {
+                    continue;
+                }
+                if (nameOfCurrentPiece == NameOfChessPiece.PAWN) {
+                    numOfPawnInFile++;
+                }
+                totalPoint += nameOfCurrentPiece.getPoint();
+            }
+            if (numOfPawnInFile >= 2) {
+                totalPoint -= numOfPawnInFile * NameOfChessPiece.SCORE_OF_SPECIAL_PAWN;
+            }
+        }
```
다음 원칙에 입각하여 설계를 다시 구상해주세요. 
이 메소드 안에서의 수정뿐만 아니라, 연관되어 있는 다른 클래스들의 수정이 필요할 수도 있습니다. 

**원칙**

- `.get` 하지 않는다. 물어본다. 
- 카운트를 올리지 않는다. _거르고 골라내서 갯수를 센다._

Malloc72P replied:
getter메서드로 ChessPiece클래스의 멤버변수값을 가져와서 비교연산하기보다, 그냥 ChessPiece객체한테 물어보는 방식으로 짜는게 더 나은 것 같네요!
```java
ChessPiece currentPiece = chessCells[rankIndex][fileIndex];
if (currentPiece.isNotSameColor(color)) {
    continue;
}
if (currentPiece.isBlank()) {
    continue;
}
```
위의 코드는 get하지 않고 물어보는 방식으로 수정한 코드예요!
조건식안에 `color != ~~` 이렇게 써두니까 확실히 이해하기가 어렵더라구요 😆 
분명히 제가 짠 코드인데도 다시보니까 `어? 이게 뭐였더라?`라고 생각하게 되네요 ㅎㅎ
그래도 바꾸고 나니까 아, 이 코드는 ChessPiece한테 아규먼트로 보낸 색이랑, 얘가 가진 색이랑 같은지 물어보는 코드구나! 하고 쉽게 알 수 있어서 좋은 것 같아요! 감사합니다!

아 그리구, `카운트를 올리지 않는다` 이 부분이 잘 이해가 안가요. `numOfPawnInFile`변수를 말씀하신 것 같은데 혹시 맞나요? 이 변수를 이용해서 카운트를 한 이유는, 세로줄에 폰이 연속해서 있으면 0.5점으로 계산해야해서 이렇게 작성했어요. 혹시 이런 방법을 선택하면 어떤 문제가 발생할 수 있을까요? 곰곰히 고민해봤는데 저는 괜찮을 것 같아서요.

wheejuni commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -108,34 +99,136 @@ private String getRepresentationOfRank(int rankIdx) {
+        return totalPoint;
+    }
+
+    public List<ChessPiece> getPiecesOfColor(ColorOfChessPiece color, boolean isAsc) {
+        List<ChessPiece> pieceList = new ArrayList<>();
+        for (int fileIndex = 0; fileIndex < LENGTH_OF_BOARD; fileIndex++) {
+            for (int rankIndex = 0; rankIndex < LENGTH_OF_BOARD; rankIndex++) {
+                ChessPiece currentPiece = chessCells[rankIndex][fileIndex];
+                if (currentPiece.getColorOfChessPiece() != color) {
+                    continue;
+                }
+                NameOfChessPiece nameOfCurrentPiece = currentPiece.getNameOfChessPiece();
+                if (nameOfCurrentPiece == NameOfChessPiece.NO_PIECE) {
+                    continue;
+                }
+                pieceList.add(currentPiece);
+            }
+        }
+        if (isAsc) {
+            pieceList.sort(createBlackBishop().comparatorAsc);
+        } else {
+            pieceList.sort(createBlackBishop().comparatorDesc);
+        }
+        return pieceList;
+    }
```
여기서도 같은 설계 원칙이 적용돼야 할 겁니다.

Malloc72P replied:
넵! 이 부분도 get을 하기보다 ChessPiece한테 물어보는 방식으로 수정했습니다!
이전보다 더 읽기 좋은 코드가 된 것 같아서 너무 좋습니다!
```java
ChessPiece currentPiece = chessCells[rankIndex][fileIndex];
if (currentPiece.isNotSameColor(color)) {
    continue;
}
if (currentPiece.isBlank()) {
    continue;
}
pieceList.add(currentPiece);

```
혹시 더 개선할 만한 부분이 있다면 알려주시면 감사하겠습니다! 😄

wheejuni commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,94 +1,85 @@
 
+import static net.nas.chess.ChessCoordinate.parseStringCoordinate;
 import static net.nas.pieces.ChessPiece.*;
 import static net.nas.utils.StringUtils.appendNewLine;
```
스태틱 임포트는 아껴 주세요. 
클래스의 이름을 통해 스태틱 메소드의 의미가 비로소 완성되는 경우가 많습니다. 스태틱 임포트하지 않는 것이 오히려 가독성에 도움이 됩니다. 기계적으로 무조건 워드 카운트를 줄이는 것이 항상 클린 코드로 가는 길은 아닙니다.

Malloc72P replied:
아하.. 이게 오히려 가독성을 해칠 수도 있겠네요...
Board클래스에서  `parseStringCoordinate` 스태틱 메서드가 나오면 어라? Board클래스에 이런 스태틱 메서드가 있었나? 라는 생각이 들 수 도 있을 것 같네요...
확실하게 `ChessCoordinate.parseStringCoordinate()`라고 써주는 편이 가독성을 높이는 측면에서 더 좋을 것 같아요!
놓친 부분을 알려주셔서 정말 감사합니다! 😄

wheejuni commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -108,34 +99,136 @@ private String getRepresentationOfRank(int rankIdx) {
+            char currChar = strRank.charAt(i);
+            NameOfChessPiece nameOfCurrPiece = NameOfChessPiece.getNameByRepresentation(String.valueOf(currChar));
+            ColorOfChessPiece colorOfCurrPiece;
+            if (currChar == '.') {
+                colorOfCurrPiece = ColorOfChessPiece.BLANK;
+            } else if (Character.isUpperCase(currChar)) {
+                colorOfCurrPiece = ColorOfChessPiece.BLACK;
+            } else if (Character.isLowerCase(currChar)) {
+                colorOfCurrPiece = ColorOfChessPiece.WHITE;
+            } else {
+                throw new InvalidParameterException("Invalid color detected! currChar : " + currChar);
+            }
```
`ColorOfChessPiece` 에 스태틱 메소드가 추가된다면, 이 부분을 매우 깔끔히 바꿀 수 있을 듯합니다.

Malloc72P replied:
리뷰어님 말을 듣고 생각해보니까, 이 기능은 추출하는게 나을 것 같다는 생각이 들었어요!
확실히 표현문자를 가지고 어떤 색상인지 알아내는 기능이니까, ColorOfChessPiece Enum에 이런 기능을 가진 스테틱 메서드를 추가하는편이 좋겠더라구요.
```java
ColorOfChessPiece.getColorByRepresentation(currChar);
add(createChessPiece(nameOfCurrPiece, colorOfCurrPiece), rankIndex, i);
```
그래서 위의 코드처럼 수정할 수 있었어요! 
앞으로 프로그래밍할때 추출할 수 있는 기능은 추출하는 연습을 하겠습니다! 💪

wheejuni reviewed (changes requested):
과제 진행하시느라 수고 많으셨습니다 👏👏
몇 가지 변경 의견을 드렸는데요, 확인해주시고 더 좋은 코드를 만들 수 있었으면 좋겠습니다!

Malloc72P committed:
축약어idx 대신, 완전한 단어인 Index로 수정하라

- 코드 가독성을 위해 축약어 사용을 피하기 위해 수정함

Malloc72P committed:
가독성을 위해 불필요한 static import를 줄여라

Malloc72P committed:
initRankByString메서드에서 문자를 읽고 체스말 색상을 알아내는 기능을 추출하라

- 해당 기능을 추출하여 ColorOfChessPiece enum에 스태틱 메서드로 추가함.

Malloc72P committed:
외부에서 어떤 연산을 위해 ChessPiece의 멤버변수를 불필요하게 얻어오지 않도록 수정해라

- ChessPiece의 멤버변수를 getter로 값을 가져와서 연산하기보다, ChessPiece에 물어봐서
  true, false여부를 획득하는 편이 코드를 이해하는것도 그렇고, 실수를 줄이는 부분에도 도움이 될 것 같아서 수정함
- 해당 수정 이후, ChessPiece에서, name과 color에 대한 getter메서드는 더이상 필요하지 않아서 제거함.

Malloc72P committed:
Board클래스에서 numberOfPieces멤버변수를 제거하라

- 효율성을 위해서 numberOfPieces멤버변수를 만들어 쓰려고 했으나,
  해당 변수로 인해 복잡도가 크게 증가하고, 오류가 발생할 가능성이 커질 것 같아서 제거하도록 수정함
- size메서드는 2차원 배열을 전체순회하면서 개수를 세도록 수정함.

Malloc72P commented:
안녕하세요?
코멘트 해주신 내용 잘 읽었어요! 😄 
리뷰내용을 반영해서 추가로 커밋했어요!

추가로 수정한 내용은 아래와 같습니다.
#### 1. 축약어 idx를 index로 전부 수정했습니다!
#### 2. static import대신, 명확하게 `ChessCoordinate.parseStringCoordinate()`로 수정했습니다!
#### 3. `initRankByString()`메서드에 있는 기능을 추출해서 `ColorOfChessPiece` enum에 메서드를 추가했습니다!
#### 4. `.get`을 쓰기 보다, ChessPiece클래스에 물어봐서 비교연산을 수행할 수 있도록 수정했습니다!
* 이렇게 수정하니까 가독성도 좋아지고, 실수할 일도 줄어들 것 같습니다!
#### 5. Board클래스에서 `numberOfPieces` 멤버변수를 제거했습니다!
* 프로그램의 복잡도를 낮추는데 도움이 될 것 같습니다!

수정한 내용은 이상입니다!
제 코드를 리뷰해주셔서 감사합니다!

wheejuni reviewed (approved):
수고 많으셨습니다 💯
